### ISSUE TITLE
Large arrays and `scan_object_and_trace_edges`

### ISSUE THREAD

**Issue by wks** [mmtk/mmtk-core#985]
TL;DR: Some arrays are very big.  If we use `scan_object_and_trace_edges` to scan them, it will take too much time in one single work packet.  We need a way to split the scanning of a large array into multiple work packets, and one way to do it is allowing enqueuing edges (or preferrably slices) in `scan_object_and_trace_edges` in order to defer the scanning of array elements.

# Problem

Ruby uses "object-enqueuing" tracing.  It enqueues objects, and uses `scan_object_and_trace_edges` to scan and object and trace its edges at the same time.

One limitation is that `scan_object_and_trace_edges` must fully scan one object and trace all of its edges before continuing.  However, one object may have many edges, and fully scanning such a large object will take a long time.  Examples include arrays, hashes, "ordinary" objects (i.e. `T_OBJECT`.  Due to the dynamic nature of Ruby, an object may have many fields added at run time.), etc.  If one `ScanObjects` work packet contains such an object, that particular work packet will take a very long time, making it a bottleneck, and reducing parallelism.  The following timeline is generated when running the GCBench benchmark.  The benchmark contains a long-lived array of 50000 elements.  In comparison, each `ScanObjects` work packets contain at most 512 elements.

![ruby-scan-object](https://github.com/mmtk/mmtk-core/assets/370317/1128360a-e336-419c-a8fd-438f407674b9)

This problem does not manifest in OpenJDK because OpenJDK uses edge-enqueuing tracing.  After scanning an `Object[]`, its individual fields are enqueued into `ProcessEdgesWork` work packets, and each work packet contains at most 4096 elements.

# Solution

Currently, the `Scanning::scan_object` method provides an `EdgeVisitor` for enqueuing edges, while `Scanning::scan_object_and_trace_edges` provides an `ObjectTracer` for tracing objects.  We need to extend `Scanning::scan_object_and_trace_edges` so that the VM can choose whether to enqueue object or enqueue edge according to the object size.

```rust
pub trait Scanning<VM: VMBinding> {
    fn scan_object_and_trace_or_enqueue_edges<EV: EdgeVisitor<VM::VMEdge>, OT: ObjectTracer>(
        tls: VMWorkerThread,
        object: ObjectReference,
        edge_visitor: &mut EV,
        object_tracer: &mut OT,
    );
}
```

The VM binding can implement this method by looking at the object type.  If it is a small object, the binding can call `object_tracer.trace_object` on its edges to trace them directly.  If it is an array or other forms of large objects, the binding can use `edge_visitor.visit_edge` to enqueue its edges so that the edges will be scanned in a dedicated `ProcessEdgesWork` work packet.  The binding can also scan some of its edges (such as the klass pointers or some header fields) directly, and enqueue other edges (such as array elements) into work packets, balancing simplicity and load-balancing.

## Merging `EdgeVisitor` and `ObjectTracer` into one trait

We can also introduce a super trait that includes both `EdgeVisitor` and `ObjectTracer`:

```rust
trait PostponableObjectTracer: ObjectTracer + EdgeVisitor {}
```

"Postponable" means it can postpone the tracing of an edge until later in a `ProcessEdgesWork` work packet.  In essence, it is still a helper object for scanning an object and tracing its edges.

# Related issues and PRs

https://github.com/mmtk/mmtk-core/pull/865 proposed adding a method `trace_object_immediately` into `EdgeVisitor`.  It intended to solve a different problems, i.e some edges (such as "offsetted" edges) are easier to trace immediately because the object's type information (such as the offsets of each edge) makes it easier to trace the edge.

I think `EdgeVisotor` with `trace_object_immediately` added is just like the `PostponableObjectTracer`, albeit described in a slightly more edge-enqueuing-centric way.

With this change, we can let the Julia binding's `Scanning::support_edge_enqueuing` return `false` so that all objects will be scanned with the `scan_object_and_trace_or_enqueue_edge` method.

*(There is one naming problem because `scan_object_and_trace_or_enqueue_edge` actually supports edge enqueuing.  Maybe we can rename ``Scanning::support_edge_enqueuing` to `Scanning::how_to_scan`, and let the VM choose between `scan_object`, `scan_object_and_trace_edges`, and `scan_object_and_trace_or_enqueue_edge`.)*

And then within `scan_object_and_trace_or_enqueue_edge`, it still enqueue edges as usual.  But when it sees `OffsetEdge`, it trace through them immediately using the `object_tracer`.

**Comment by wks** [comment]
It is unconventional for a GC implementation to support both slot (edge) enqueuing and object enqueuing.  Traditionally, a GC algorithm uses either slot (edge) enqueuing or object enqueuing.  But this doesn't mean the idea described in this issue is not worth doing in our current implementation.

We may reconsider the design of object/edge enqueuing.  It is better to do so with our researches on prefetching taken into account.
